Added unit conversion capability

zhieejhia93 opened the pull request:
This PR adds the unit-conversion capability to WATTS:

- `params` can now accept dictionary where the `value`, the `current_unit`, and the `new_unit` of a parameter are specified. 
- `new_unit` is optional. If not specified, S.I. units will be used.
- Changes are also made to MOOSE and OpenMC plugins to automatically convert the units of a parameter to S.I and CGS, respectively. With the changes, there is no need to define the same parameter twice in different units to accommodate for the different units requirements in SAM/MOOSE and OpenMC.

zhieejhia93 committed:
Added unit conversion capability

zhieejhia93 commented:
@nstauff @paulromano Please have a look. Thank you.

nstauff commented on src/watts/plugin_moose.py:
```diff
@@ -167,8 +167,21 @@ def prerun(self, params: Parameters):
+                    params[key] = params.convert_units(dict_param=dict_param)
+                    params.converted_param.append(key)
+                else:
+                    params[key] = params.convert_units(dict_param=dict_param, system='si')
```
This new section seems to be duplicated in different plugin - would it make sense to push it as a new procedure at a higher level (`plugin.py`) that takes `si` or `cgs` as an entry - this would enable it for any new application.

nstauff commented on examples/example2_SAM_OpenMC/example2.py:
```diff
@@ -59,18 +59,18 @@
+    print(key, moose_result.csv_data[key])
+print(moose_result.inputs)
+print(moose_result.outputs)
 
```
thanks for making this update. I thought I had done it... Could you also make this change for `example3` and `example4`?

nstauff commented:
@zhieejhia93 thanks for this work! I like the approach. Could you please update the documentation in `usage.rst`?

paulromano commented on examples/example1a_SAM/example1a.py:
```diff
@@ -9,13 +9,13 @@
 # TH params
 
-params['He_inlet_temp'] = 600 + 273.15  # K
+params['He_inlet_temp'] = {"value":600, "current_unit": "Celsius", "new_unit": "Kelvin"}  #873.15 K
```
In these cases where we know what the final unit is, I don't see how we're providing any useful functionality versus what the underlying library provides. For example, this could just as well be:
```python
from astropy import units as u
...
params['He_inlet_temp'] = (600.0 * u.deg_C).to_value('K', equivalencies=u.temperature())
```

paulromano commented on setup.cfg:
```diff
@@ -35,6 +35,7 @@ install_requires =
     pandas
     platformdirs
     prettytable
+    astropy
```
Can you elaborate a little bit on why you chose astropy.units as opposed to some of the other libraries available?

zhieejhia93 replied:
I looked at different libraries, it seemed to me that Pint, unyt, and astropy were more robust than the other libraries. At the end I chose astropy because it was the easiest to convert between different unit systems when `new_unit` is not specified. With other libraries, the conversion between different systems seems to require the user to specify 'new_unit`.

paulromano commented on src/watts/parameters.py:
```diff
@@ -289,3 +295,44 @@ def from_hdf5(cls, filename_or_obj: Union[str, h5py.Group]):
+                    convert_value = current_value.cgs
+                elif system == 'si':
+                    convert_value = current_value.si
+        return convert_value.value 
```
Right now this method doesn't actually use `self` at all, which begs the question of whether it should just be a normal function instead

paulromano commented on src/watts/plugin_moose.py:
```diff
@@ -167,8 +167,21 @@ def prerun(self, params: Parameters):
+                dict_param = params.dict_unit_conv[key]
+
+                if "new_unit" in dict_param.keys():
+                    params[key] = params.convert_units(dict_param=dict_param)
```
One thing I think we can improve here is that the plugin classes are actually modifying the `params` that are getting passed in -- ideally that argument should be unchanged. An easy way around this would be for the plugin to make a copy of the `Parameters` instance before making any changes in it.

paulromano reviewed (changes requested):
Thanks for the first cut at this @zhieejhia93. I think we can improve upon this a bit -- right now, the implementation relies on a dictionary, but we also allow users to store information in dictionaries without them having a special meaning. I'm thinking that instead, we can have the `Parameters` class just store instances of `astropy.units.Quantity`. At the time a plugin class is about to write inputs, it can take the `Parameters` and convert any stored `Quantity` instances into normal numbers (based on the preferred unit system of the plugin). Perhaps it's easier to work through this in a meeting together. Let me know if you'd like to set something up for next week.

Besides the design changes, another thing we'll need before this is ready to be merged is a set of unit tests to confirm that it works as expected.

zhieejhia93 commented:
Thanks for the comments @nstauff @paulromano . I will try to implement the suggestions by @paulromano and let you both know if I have any questions.

zhieejhia93 committed:
Merge branch 'watts-dev:development' into devel-unit-conversion

zhieejhia93 committed:
Modified unit conversion capability

zhieejhia93 committed:
Merge branch 'devel-unit-conversion' of https://github.com/zhieejhia93/watts into devel-unit-conversion

Pull changes

zhieejhia93 commented:
Hi @nstauff  @paulromano I made some changes based on your comments:

- The implementation still relies on dictionary to allow users to input the `value` and `current_unit` of a parameter. I removed the capability to specify `new_unit` because as Paul mentioned previously, it didn't add much value to the code as users could simply perform the unit conversion in the Python input file.
- The parameter that needs to be converted is now stored as `astropy.units.Quantity` in the `Parameter` class and is converted to SI or CGS depending on whether it's being used by MOOSE or openmc. 
- As Paul mentioned previously, ideally `params` should remain unchanged. In this implementation, unit conversion is performed on a copy of `params`, which is then passed to the template generator.
- I kept the unit conversion at the MOOSE and OpenMC plugins level instead of the `plugin` level as suggested by Nicolas because I wasn't sure if that was possible without changing `params`.
- I added a unit test to `test_results.py`. The unit test simply checks whether the unit conversion is performed correctly, I am not sure if that is sufficient or more checks are required.
- I have been testing the implementation with example1a and example2. I will update the remaining example files once this is finalized.

Thank you.

nstauff commented:
Thanks @zhieejhia93! my main comment to push some of the logic to the `plugin` level (to avoid duplication in different plugins) was not addressed - maybe @paulromano can help with this improvement.
Otherwise this looks good to me - but I will let @paulromano do a detailed review.

paulromano commented on examples/example1a_SAM/example1a.py:
```diff
@@ -45,10 +45,10 @@
 moose_app_type = "SAM"
 app_dir = os.environ[moose_app_type.upper() + "_DIR"]
-moose_plugin = watts.PluginMOOSE(moose_app_type.lower() + '_template', show_stdout=True) # show all the output
+moose_plugin = watts.PluginMOOSE(moose_app_type.lower() + '_template', show_stdout=False) # show all the output
```
```suggestion
moose_plugin = watts.PluginMOOSE(moose_app_type.lower() + '_template')
```

paulromano commented on examples/example1a_SAM/example1a.py:
```diff
@@ -9,13 +9,13 @@
-params['He_inlet_temp'] = 600 + 273.15  # K
-params['He_outlet_temp'] = 850 + 273.15 # K
-params['He_cp'] = 5189.2 # J/kg-K
+params['He_inlet_temp'] = {"value":600, "current_unit": "Celsius"}  # 873.15 K
```
I'm still not a fan of providing this as a dictionary, since we already allow that for other purposes. I would suggest just setting this using the `Quantity` class directly:
```Python
from astropy.units import Quantity
...

params['He_inlet_temp'] = Quantity(600, "Celsius")
```

paulromano commented on src/watts/plugin_moose.py:
```diff
@@ -205,3 +211,24 @@ def postrun(self, params: Parameters) -> ResultsMOOSE:
+        self.params_copy = copy.deepcopy(params)
+
+        for key in self.params_copy.keys():
+            if key in self.params_copy.convert_param:
```
Rather than keeping track of `convert_param`, you can just have a check on the type of the value:
```Python
if isinstance(self.params_copy[key], astropy.units.Quantity):
    ...
```

paulromano commented on src/watts/plugin_moose.py:
```diff
@@ -205,3 +211,24 @@ def postrun(self, params: Parameters) -> ResultsMOOSE:
+    def _convert_unit(self, params: Parameters):
+        # Function to convert units to SI
+        u.imperial.enable()
+        self.params_copy = copy.deepcopy(params)
```
Rather than setting a new attribute, just have this method return the new copy of params. Then in prerun, you'd have:
```
params_copy = self._convert_unit(params)
...
super().prerun(params_copy, ...)

paulromano commented on tests/test_results.py:
```diff
@@ -6,6 +6,7 @@
 
 import watts
 import numpy as np
+import os
```
```suggestion
```

paulromano commented on tests/test_results.py:
```diff
@@ -103,3 +104,27 @@ def test_results_moose(run_in_tmpdir):
+
+def test_results_unit_conversion(run_in_tmpdir):
+    # Create a fake template file
+    file = open("sam_template","w")
```
```suggestion
    file = open("sam_template", "w")
```

paulromano commented on src/watts/plugin_moose.py:
```diff
@@ -205,3 +211,24 @@ def postrun(self, params: Parameters) -> ResultsMOOSE:
+                # astropy uses a different method to convert temperature.
+                # Variables are converted to SI by default.
+
+                if self.params_copy[key].unit in temperature_units:
+                    self.params_copy[key] = self.params_copy[key].to(u.K, equivalencies=u.temperature()).value
+                else:
+                    self.params_copy[key] = self.params_copy[key].si.value
```
I agree with @nstauff that we can push this into the general `Plugin` class. To have plugin-specific unit systems, this could be changed to something like:
```Python
if self.params_copy[key].unit in temperature_units:
    self.params_copy[key] = self.params_copy[key].to(self.unit_temperature, equivalences=u.temperature()).value
else:
    self.params_copy[key] = getattr(self.params_copy[key], self.unit_system).value
```

and then in the plugins themselves:
```Python
class PluginMOOSE(TemplatePlugin)
    """..."""

    unit_system = 'si'
    unit_temperature = 'K'

    def __init__(self, ...):
```
and
```Python
class PluginOpenMC(Plugin):
    """..."""

    unit_system = 'cgs'
    unit_temperature = 'K'

    def __init__(self, ...):
```

paulromano reviewed (changes requested):
Thanks for the updates @zhieejhia93. Here is some more feedback -- let me know if you have questions or you prefer to setup a call to walk through some of these.

zhieejhia93 commented:
Thanks @nstauff  and @paulromano. I will work on the comments and let you both know if I have any questions.

zhieejhia93 committed:
Merge branch 'watts-dev:development' into devel-unit-conversion

zhieejhia93 committed:
Moved unit conversion to plugin level

zhieejhia93 commented on src/watts/plugin_openmc.py:
```diff
@@ -118,6 +116,20 @@ def __init__(self, model_builder: Callable[[Parameters], None],
         self.show_stdout = show_stdout
         self.show_stderr = show_stderr
 
+    def convert_unit(self, params: Parameters, unit_system, unit_temperature):
```
Added this method explicitly to allow the unit test to check that the unit conversion is correct.

zhieejhia93 commented on src/watts/plugin_moose.py:
```diff
@@ -162,6 +161,20 @@ def options(self, moose_exec):
         """
         self.moose_exec = moose_exec
 
+    def convert_unit(self, params: Parameters, unit_system, unit_temperature):
```
Added this method explicitly to allow the unit test to check that the unit conversion is correct.

paulromano replied:
The method as defined in `Plugin` will be available by inheritance, so it's not necessary to explicitly define it in PluginMOOSE and PluginOpenMC

zhieejhia93 commented:
Hi @nstauff  @paulromano, I modified the implementation according to your comments:

- The unit conversion implementation has been moved to the general `plugin` level as recommended.
- Removed the dictionary input style. Users can now use the `Quantity` class from Astropy directly to input the value and units of parameters.
- Updated the unit test to check the unit conversion in both MOOSE and OpenMC.

Thank you

zhieejhia93 commented on src/watts/plugin.py:
```diff
@@ -84,6 +86,45 @@ def workflow(self, params: Parameters, name='Workflow') -> Results:
+                if params_copy[key].unit in temperature_units:
+                    params_copy[key] = params_copy[key].to(u.Unit(unit_temperature), equivalencies=u.temperature()).value
+                else:
+                    if unit_system == 'cgs':
```
@paulromano I tried to implement the method that you suggested but couldn't get it to work. As a temporary solution, I decided to set `unit_temperature` and `unit_system` as arguments and use an `if-else` statement as a workaround. Please let me know how can I fix this. Thanks.

paulromano replied:
I think this is actually OK for now

paulromano commented on src/watts/plugin.py:
```diff
@@ -84,6 +86,45 @@ def workflow(self, params: Parameters, name='Workflow') -> Results:
 
         return result
 
+    def convert_unit(self, params: Parameters, unit_system, unit_temperature):
```
```suggestion
    def convert_unit(self, params: Parameters, unit_system: str, unit_temperature: str):
```

paulromano commented on src/watts/plugin.py:
```diff
@@ -84,6 +86,45 @@ def workflow(self, params: Parameters, name='Workflow') -> Results:
+                # Variables are converted to SI by default.
+
+                if params_copy[key].unit in temperature_units:
+                    params_copy[key] = params_copy[key].to(u.Unit(unit_temperature), equivalencies=u.temperature()).value
```
Isn't it sufficient to just pass a string as the first argument to the `to()` method? i.e.
```suggestion
                    params_copy[key] = params_copy[key].to(unit_temperature, equivalencies=u.temperature()).value
```

paulromano commented on src/watts/plugin.py:
```diff
@@ -84,6 +86,45 @@ def workflow(self, params: Parameters, name='Workflow') -> Results:
+                        params_copy[key] = params_copy[key].cgs.value
+                    elif unit_system == 'si':
+                        params_copy[key] = params_copy[key].si.value
+        return(params_copy)
```
```suggestion
        return params_copy
```

paulromano commented:
@zhieejhia93 It looks like CI tests are failing due to an update of the OpenMC docker image -- don't worry about that, I can fix it in a follow-on PR

zhieejhia93 committed:
Minor changes to tidy up the code

zhieejhia93 commented:
@paulromano I made the changes that you suggested. Please have a look. Also, do I have to do anything on my end regarding the failed CI tests? Thanks.

paulromano reviewed (approved):
Looks good to me now. Thanks for bringing this to the finish line! The one remaining issue is that writing parameters to HDF5 and then reading back will not preserve `Quantity` instances, but that will require some more extensive changes in the HDF5 handling, so I'll take a shot at fixing that with a follow-on PR.

You don't need to do anything about the CI tests -- once #8 is merged it should be fine.

zhieejhia93 commented:
@nstauff @paulromano Thanks for the feedback and comments on this PR.

nstauff commented:
@zhieejhia93 thanks for your work on this PR! This looks good to me. A few additional things that will need addressing in future PR:
- documentation for future users
- extend this to the new PyARC plugin

zhieejhia93 commented:
@nstauff Sure, I can do that. Just to double check, does PyARC use SI units as default?
